### PR TITLE
refactor path resolving

### PR DESCRIPTION
**Pull request by xvuko**
Main point of this is to remove `char tmp_buf[PATH_MAX];` from stack. I refactored rest of the function to make it (hopefully) easier to read. 

This should give same output as before. Only change is case of empty symlink as it was setting errno to 0.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [ ] Tested by hand on: (list targets here).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [ ] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

**Review by niewim19 — COMMENTED** [review]
Hey, I have done some research on this change and I found some issues within this change and in current implementation. @damianloew I have described current issue in section 2 of this comment. I will also make an issue on this.

# Section 1, issue in proposed change

Using `readlink` instead of `stat` looks neat, considering how many messages are sent by `stat` (~10). Proposed change uses spare place in `path` argument to expand and process symlinks content. This may be erroneous if original path is sufficiently long so that not all data read from symlink can be stored. Such clipped symlink (`readlink` does not fail if place to write is too small) content can be data-wise valid and resolve to _some place_ in filesystem. It would be a byzantine error, and very hard to find.

Test code below works on host and current implementation but causes segmentation fault on proposed change.

# Section 2, error in current implementation

Makeshift code that I have produced to check Section 1 error has found a discrepancy between current implementation of `resolve_path()` (via `realpath()`) and the host one.

Test code below resolves to `/etc` on host, but `/symShort` on phoenix, which is in violation with goal of `realpath()` which should remove symlinks.

# Test code

Test code tries to resolve very long path that starts with symlink. It instantiates:
 - symlink `symNameLong` with long name that points to "/"
 - symlink `symNameShort` that points to `symNameLong`
 - `basePath` path that goes like: `"symShort/etc/../etc/<and so on up to almost PATH_MAX length>/../etc"` that should finally land in `etc`.

```c
#include <stdio.h>
#include <unistd.h>
#include <sys/stat.h>
#include <stdlib.h>
#include <limits.h>
#include <string.h>
#include <errno.h>

#define NAME_LEN (NAME_MAX / 2)

int main(int argc, char **argv) {
	char symNameShort[] = "symShort";
	char symNameLong[NAME_LEN];
	char basePath[] = "symShort/etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/../etc/";
	char resolved[PATH_MAX + 1];

	memset(symNameLong, 'a', NAME_LEN);
	symNameLong[NAME_LEN - 1] = 0;

	if (symlink("/", symNameLong) < 0)
		printf("%s\n", strerror(errno));

	if (symlink(symNameLong, symNameShort) < 0)
		printf("%s\n", strerror(errno));

	realpath(basePath, resolved);
	resolved[PATH_MAX] = 0;
	printf("%s\n", resolved);

	/* cleanup */
	remove(symNameShort);
	remove(symNameLong);
}
```

`basePath` is resolved to:
 - `/etc` on host
 - `/symShort` on phoenix (tested on zturn)
 - `segmentation fault` on proposed change

**Comment by nalajcie** [comment]
small notes from me:
 - the code looks sane (but I haven't tested it), I was thinking about similar approach at the initial approach but decided the increased code complexity wasn't worth the case
 - please write in commit message body regarding WHY this change is needed (or what we gained) to leave trail
 - @niewim19 the `_readlink_abs` instead of `lstat` was in the original approach so this is not a functionality change
 - if there is a test that highlights possible error in current / new implementation - please try to add it to unit tests :)
 - I'll try to test it in the common use cases on DTR project, will follow up then

File: unistd/dir.c [hunk 1]
```diff
@@ -154,111 +154,109 @@ char *canonicalize_file_name(const char *path)
...
-			is_leaf = 1; /* last path node */
-		}
-		else {
-			char *rem = slash;
-			while (*rem == '/')
-				rem++;
-			if (*rem == 0)
-				is_leaf = 1; /* trailing '/' */
+	while (*p != '\0') {
+		char *node = p;
```
Comment (nalajcie) [inline unistd/dir.c:172]: `const`?

File: unistd/dir.c [hunk 2]
```diff
@@ -154,111 +154,109 @@ char *canonicalize_file_name(const char *path)
...
+		p = strchrnul(node, '/');
+		size_t node_len = p - node;
```
Comment (nalajcie) [inline unistd/dir.c:174]: `const`?

**Comment by niewim19** [comment]
>     * @niewim19 the `_readlink_abs` instead of `lstat` was in the original approach so this is not a functionality change
> 
>     * if there is a test that highlights possible error in current / new implementation - please try to add it to unit tests :)
> 
>     * I'll try to test it in the common use cases on DTR project, will follow up then

- I`m sorry for that - it is indeed not a functionality change. I have misunderstood diff. I will amend previous message
- I have spoke with test team (@damianloew) about this issue, they will take it from here.

**Comment by damianloew** [comment]
> > small notes from me:
> > ```
> > * the code looks sane (but I haven't tested it), I was thinking about similar approach at the initial approach but decided the increased code complexity wasn't worth the case
> > 
> > * please write in commit message body regarding WHY this change is needed (or what we gained) to leave trail
> > 
> > * @niewim19 the `_readlink_abs` instead of `lstat` was in the original approach so this is not a functionality change
> > 
> > * if there is a test that highlights possible error in current / new implementation - please try to add it to unit tests :)
> > 
> > * I'll try to test it in the common use cases on DTR project, will follow up then
> > ```
> 
> * I`m sorry for that - it is indeed not a functionality change. I have misunderstood diff. I will amend previous message
> * I have spoke with test team (@damianloew) about this issue, they will take it from here.

Yeah, the issue will be created and we will add tests to cover this issue.

**Comment by agkaminski** [comment]
@xvuko Hmmm, I don't know... Doesn't this solution cause us to be unable to on a long path not being able to solve even longer symlink? Total path length being shorter than PATH_MAX in both cases, but us being unable to solve it anyway

**Comment by nalajcie** [comment]
> @xvuko Hmmm, I don't know... Doesn't this solution cause us to be unable to on a long path not being able to solve even longer symlink? Total path length being shorter than PATH_MAX in both cases, but us being unable to solve it anyway

I think the original solution would also fail if symlink resolved length + remaining (unresolved) path len exceeds PATH_MAX (while resolving, even if the expected resolve result would be < PATH_MAX).

**Comment by damianloew** [comment]
> > @xvuko Hmmm, I don't know... Doesn't this solution cause us to be unable to on a long path not being able to solve even longer symlink? Total path length being shorter than PATH_MAX in both cases, but us being unable to solve it anyway
> 
> I think the original solution would also fail if symlink resolved length + remaining (unresolved) path len exceeds PATH_MAX (while resolving, even if the expected resolve result would be < PATH_MAX).

We checked it and in fact there is no issue in the current libphoenix implementation as you mentioned @nalajcie So I've updated the issue and closed it. However the exception shouldn't be raised when applying changes from the PR when running code from the mentioned case.

**Comment by xvuko** [comment]
>     * `segmentation fault` on proposed change

Thanks for testing this. There was a bug in my code:
```c
assert(symlink_len <= path - p);                                                                                                                                                                                                                                                                                                                                          
if (symlink_len == path - p) {                                                                                                                                                                                                                                                                                                                                            
      return SET_ERRNO(-ENAMETOOLONG);                                                                                                                                                                                                                                                                                                                                  
}
```
I changed `path - p` to `p - path` and now `/symShort` is printed.

**Comment by nalajcie** [comment]
- ok, so maybe adding the mentioned test would still be beneficial?
- we could introduce `symlink_max_len = p - path`  but I'm not sure if it would increase readability or not
- (Just analyzing code) I see that `/` is being added unconditionally (also if leaf) so we might end up in with eg. `/dir/realfile/` in `path` - but from reading the code it seems it would correctly be handled? (this reduces max resolved symlink len by one from the previous implementation but IMHO doesn't really matter)

**Comment by xvuko** [comment]
> * ok, so maybe adding the mentioned test would still be beneficial?
Yes, I will add this test.

>     * we could introduce `symlink_max_len = p - path`  but I'm not sure if it would increase readability or not
Yeah, I was thinking about that. Added as it is a good place for comment.

>     * (Just analyzing code) I see that `/` is being added unconditionally (also if leaf) so we might end up in with eg. `/dir/realfile/` in `path` - but from reading the code it seems it would correctly be handled? (this reduces max resolved symlink len by one from the previous implementation but IMHO doesn't really matter)

'/' at the and is not an issue as trailing slashes are always ignored.

File: unistd/dir.c [hunk 3]
```diff
@@ -154,111 +154,115 @@ char *canonicalize_file_name(const char *path)
...
-				return SET_ERRNO(-ENAMETOOLONG);
+		if (r + 1 + node_len + 1 > result + PATH_MAX)
+			return SET_ERRNO(-ENAMETOOLONG);
+
+		*r++ = '/';
+		memcpy(r, node, node_len);
+		r += node_len;
+		*r = '\0';
+
+		if (is_leaf && !resolve_last_symlink) {
```
Comment (agkaminski) [inline unistd/dir.c:204]:
```suggestion
		if ((is_leaf != 0) && (resolve_last_symlink == 0)) {
```

File: unistd/dir.c [hunk 4]
```diff
@@ -154,111 +154,115 @@ char *canonicalize_file_name(const char *path)
...
 		}
-		else { /* 'normal' path node */
-			char tmp_buf[PATH_MAX];
-			ssize_t symlink_len;
 
-			/* check for buffer overflows */
-			if (r + (slash - p + 1) > res_end)
-				return SET_ERRNO(-ENAMETOOLONG);
+		if (r + 1 + node_len + 1 > result + PATH_MAX)
+			return SET_ERRNO(-ENAMETOOLONG);
```
Comment (agkaminski) [inline unistd/dir.c:197]:
```suggestion
		if ((r + 1 + node_len + 1) > (result + PATH_MAX)) {
			return SET_ERRNO(-ENAMETOOLONG);
		}
```

File: unistd/dir.c [hunk 5]
```diff
@@ -154,111 +154,115 @@ char *canonicalize_file_name(const char *path)
...
+			continue;
 		}
-		else if (p[0] == '.' && p[1] == 0) { /* '.' dir */
-			r -= 1;
+		if (node_len == 1 && node[0] == '.') { /* '.' dir */
+			continue;
 		}
-		else if (p[0] == '.' && p[1] == '.' && p[2] == 0) { /* '..' dir */
-			r -= 1;
+		if (node_len == 2 && node[0] == '.' && node[1] == '.') { /* '..' dir */
```
Comment (agkaminski) [inline unistd/dir.c:190]:
```suggestion
		if ((node_len == 2) && (node[0] == '.') && (node[1] == '.')) { /* '..' dir */
```

File: unistd/dir.c [hunk 6]
```diff
@@ -154,111 +154,115 @@ char *canonicalize_file_name(const char *path)
...
-			else if ((symlink_len = _readlink_abs(result, tmp_buf, sizeof(tmp_buf) - 1)) < 0) { /* reserve space for \0 */
-
-				if ((errno == EINVAL) || ((errno == ENOENT) && is_leaf && allow_missing_leaf)) { /* not a link or non-esixting leaf */
-					r += slash - p;
-				}
-				else {
-					/* errno set by _readlink_abs */
-					return -1;
-				}
+			else if ((errno == ENOENT) && is_leaf && allow_missing_leaf) { /* non-esixting leaf */
```
Comment (agkaminski) [inline unistd/dir.c:222]:
```suggestion
			else if ((errno == ENOENT) && (is_leaf != 0) && (allow_missing_leaf != 0)) { /* non-esixting leaf */
```

File: unistd/dir.c [hunk 7]
```diff
@@ -154,111 +154,115 @@ char *canonicalize_file_name(const char *path)
...
-			*r++ = '/';
-			p = slash + 1;
+		if (symlink_len == 0) {
+			/* NOTE: this case is not defined in POSIX */
+			/* see: https://lwn.net/Articles/551224/ */
+			return SET_ERRNO(-ENOENT);
+		}
+
+		if (++symlinkcnt > SYMLOOP_MAX)
+			return SET_ERRNO(-ELOOP);
```
Comment (agkaminski) [inline unistd/dir.c:238]:
```suggestion
		++symlinkcnt;
		if (symlinkcnt > SYMLOOP_MAX) {
			return SET_ERRNO(-ELOOP);
		}
```

**Review by agkaminski — COMMENTED** [review]
Some MISRA-flavored nitpicks, otherwise LGTM

**Comment by damianloew** [comment]
> > * ok, so maybe adding the mentioned test would still be beneficial?
> 
> Yes, I will add this test.
> 
> > ```
> > * we could introduce `symlink_max_len = p - path`  but I'm not sure if it would increase readability or not
> > ```
> 
> Yeah, I was thinking about that. Added as it is a good place for comment.
> 
> > ```
> > * (Just analyzing code) I see that `/` is being added unconditionally (also if leaf) so we might end up in with eg. `/dir/realfile/` in `path` - but from reading the code it seems it would correctly be handled? (this reduces max resolved symlink len by one from the previous implementation but IMHO doesn't really matter)
> > ```
> 
> '/' at the and is not an issue as trailing slashes are always ignored.

The test is already added and it's now running during CI.

**Review by agkaminski — APPROVED** [review]
LGTM

**Pull request merged**